File: packages/cli-py/planforge/utils/active_plan.py
"""Resolve which plan file to use for implement: index.json activePlan or latest .plan.md by mtime."""

import json
from pathlib import Path

from planforge.utils.paths import get_plans_dir

INDEX_JSON = "index.json"


def _collect_plan_files(plans_dir: Path) -> list[Path]:
    return [path for path in plans_dir.rglob("*.plan.md") if path.is_file()]


def _resolve_active_plan_candidate(plans_dir: Path, name: str) -> Path | None:
    normalized = Path(name.replace("\\", "/"))
    candidate_with_ext = (
        normalized if normalized.as_posix().endswith(".plan.md") else Path(f"{normalized.as_posix()}.plan.md")
    )
    candidates = [plans_dir / normalized, plans_dir / candidate_with_ext]
    for candidate in candidates:
        if candidate.exists():
            return candidate

    target_names = {candidate.name for candidate in candidates}
    for path in _collect_plan_files(plans_dir):
        if path.name in target_names:
            return path
    return None


def get_active_plan_path(project_root: str) -> str | None:
    plans_dir = get_plans_dir(project_root)
    pdir = Path(plans_dir)
    if not pdir.exists():
        return None
    index_file = pdir / INDEX_JSON
    if index_file.exists():
        try:
            data = json.loads(index_file.read_text(encoding="utf-8"))
            name = (data.get("activePlan") or "").strip()
            if name:
                candidate = _resolve_active_plan_candidate(pdir, name)
                if candidate:
                    return str(candidate)
        except (json.JSONDecodeError, OSError):
            pass
    latest_path = None
    latest_mtime = 0.0
    for entry in _collect_plan_files(pdir):
        mtime = entry.stat().st_mtime
        if mtime > latest_mtime:
            latest_mtime = mtime
            latest_path = str(entry)
    return latest_path

File: packages/cli-js/src/utils/active-plan.ts
/**
 * Resolve which plan file to use for implement: index.json activePlan or latest .plan.md by mtime.
 */

import { existsSync, statSync, readFileSync, readdirSync } from "fs";
import { resolve, join } from "path";
import { getPlansDir } from "./paths.js";

const INDEX_JSON = "index.json";

function collectPlanFiles(plansDir: string): string[] {
  const out: string[] = [];
  const stack = [plansDir];
  while (stack.length > 0) {
    const dir = stack.pop() as string;
    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = join(dir, entry.name);
      if (entry.isDirectory()) {
        stack.push(full);
      } else if (entry.isFile() && entry.name.endsWith(".plan.md")) {
        out.push(full);
      }
    }
  }
  return out;
}

function resolveActivePlanCandidate(plansDir: string, name: string): string | null {
  const normalized = name.replace(/\\/g, "/");
  const directCandidates = [
    resolve(plansDir, normalized),
    resolve(plansDir, normalized.endsWith(".plan.md") ? normalized : `${normalized}.plan.md`),
  ];
  for (const candidate of directCandidates) {
    if (existsSync(candidate)) return candidate;
  }

  const targetBasenames = new Set(directCandidates.map((candidate) => candidate.split(/[/\\]/).pop() ?? ""));
  for (const path of collectPlanFiles(plansDir)) {
    const basename = path.split(/[/\\]/).pop() ?? "";
    if (targetBasenames.has(basename)) return path;
  }
  return null;
}

export function getActivePlanPath(projectRoot: string): string | null {
  const plansDir = getPlansDir(projectRoot);
  if (!existsSync(plansDir)) return null;

  const indexPath = join(plansDir, INDEX_JSON);
  if (existsSync(indexPath)) {
    try {
      const raw = readFileSync(indexPath, "utf-8");
      const data = JSON.parse(raw) as { activePlan?: string };
      const name = data.activePlan?.trim();
      if (name) {
        const candidate = resolveActivePlanCandidate(plansDir, name);
        if (candidate) return candidate;
      }
    } catch {
      /* ignore invalid index.json */
    }
  }

  let latestPath: string | null = null;
  let latestMtime = 0;
  try {
    for (const full of collectPlanFiles(plansDir)) {
      const mtime = statSync(full).mtimeMs;
      if (mtime > latestMtime) {
        latestMtime = mtime;
        latestPath = full;
      }
    }
  } catch {
    return null;
  }
  return latestPath;
}
